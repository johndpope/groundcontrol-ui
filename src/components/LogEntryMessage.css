.LogEntryMessage a {
  color: inherit !important;
  text-decoration: underline;
}

.LogEntryMessage .ansi-black {
  color: #333;
}

.LogEntryMessage .ansi-red {
  color: #db2828;
}

.LogEntryMessage .ansi-green {
  color: #21ba45;
}

.LogEntryMessage .ansi-yellow {
  color: #fbbd08;
}

.LogEntryMessage .ansi-blue {
  color: #2185d0;
}

.LogEntryMessage .ansi-magenta {
  color: #e03997;
}

.LogEntryMessage .ansi-cyan {
  color: #00b5ad;
}

.LogEntryMessage .ansi-white {
  color: #fff;
}

.LogEntryMessage .ansi-bright-black {
  background-color: #333;
}

.LogEntryMessage .ansi-bright-red {
  background-color: #db2828;
}

.LogEntryMessage .ansi-bright-green {
  background-color: #21ba45;
}

.LogEntryMessage .ansi-bright-yellow {
  background-color: #fbbd08;
}

.LogEntryMessage .ansi-bright-blue {
  background-color: #2185d0;
}

.LogEntryMessage .ansi-bright-magenta {
  background-color: #e03997;
}

.LogEntryMessage .ansi-bright-cyan {
  background-color: #00b5ad;
}

.LogEntryMessage .ansi-bright-white {
  background-color: #fff;
}