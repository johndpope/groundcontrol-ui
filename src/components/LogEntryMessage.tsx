// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Ansi from "ansi-to-react";
import graphql from "babel-plugin-relay/macro";
import React from "react";
import ReactDOMServer from "react-dom/server";

import { createFragmentContainer } from "react-relay";

import { LogEntryMessage_item } from "./__generated__/LogEntryMessage_item.graphql";

import "./LogEntryMessage.css";

interface IAnchor {
  href: string;
}

function isAnchor(el: any): el is IAnchor {
  return "href" in el;
}

export interface IProps {
  item: LogEntryMessage_item;
  onClickSourceFile: (values: IProps) => any;
}

export function LogEntryMessage(props: IProps) {
  const {
    item: {
      message,
      sourceFile,
      sourceFileBegin,
      sourceFileEnd,
    },
    onClickSourceFile,
  } = props;

  if (!sourceFile) {
    return (
      <div className="LogEntryMessage">
        <Ansi
          linkify={true}
          useClasses={true}
        >
          {message}
        </Ansi>
      </div>
    );
  }

  // This is a hackish way to deal with both Ansi escape sequences and links
  // to source files, but a proper implementation will take time. Until then,
  // while far from beautiful, this should do the trick.

  const text = ReactDOMServer.renderToStaticMarkup(
    <Ansi
      linkify={true}
      useClasses={true}
    >
      {message}
    </Ansi>,
  );

  const source = message.substring(sourceFileBegin!, sourceFileEnd!);
  const href = `file://${sourceFile}`;
  const link = `<a href="${href}">${source}</a>`;
  const html = text.replace(source, link);
  const handleClick = (event: React.MouseEvent) => {
    event.preventDefault();
    if (isAnchor(event.target) && event.target.href === href ) {
      onClickSourceFile({ ...props });
    }
  };

  return (
    <div
      className="LogEntryMessage"
      dangerouslySetInnerHTML={{ __html: html }}
      onClick={handleClick}
    />
  );
}

export default createFragmentContainer(LogEntryMessage, graphql`
  fragment LogEntryMessage_item on LogEntry {
    message
    sourceFile
    sourceFileBegin
    sourceFileEnd
  }`,
);
