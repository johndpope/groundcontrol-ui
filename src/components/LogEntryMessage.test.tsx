// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { shallow } from "enzyme";
import React from "react";
import { mocked } from "ts-jest/utils";

import { mockQueryPropAttrs } from "../testing/relay";

import { LogEntryMessage } from "./LogEntryMessage";

const propsWithoutSourceFile = {
  item: {
    ...mockQueryPropAttrs(),
    message: "message",
    sourceFile: null,
    sourceFileBegin: null,
    sourceFileEnd: null,
  },
  onClickSourceFile: jest.fn(),
};

const propsWithSourceFile = {
  item: {
    ...mockQueryPropAttrs(),
    message: "message (message.ts:10)",
    sourceFile: "/path/to/message.ts:10",
    sourceFileBegin: 9,
    sourceFileEnd: 22,
  },
  onClickSourceFile: jest.fn(),
};

beforeEach(() => {
  mocked(propsWithSourceFile.onClickSourceFile).mockClear();
});

describe("<LogEntryMessage />", () => {

  it("renders correctly without a source file", () => {
    const wrapper = shallow(<LogEntryMessage {...propsWithoutSourceFile} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly with a source file", () => {
    const wrapper = shallow(<LogEntryMessage {...propsWithSourceFile} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("triggers onClickSourceFile when the source file is clicked", () => {
    const wrapper = shallow(<LogEntryMessage {...propsWithSourceFile} />);
    const event = { preventDefault: jest.fn(), target: { href: "file:///path/to/message.ts:10" } };
    wrapper.find("div").simulate("click", event);
    expect(event.preventDefault).toBeCalledTimes(1);
    expect(propsWithSourceFile.onClickSourceFile).toBeCalledTimes(1);
    expect(propsWithSourceFile.onClickSourceFile).toBeCalledWith(propsWithSourceFile);
  });

});
