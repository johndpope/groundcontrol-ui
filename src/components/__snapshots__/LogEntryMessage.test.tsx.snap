// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LogEntryMessage /> renders correctly with a source file 1`] = `
<div
  className="LogEntryMessage"
  dangerouslySetInnerHTML={
    Object {
      "__html": "message (<a href=\\"file:///path/to/message.ts:10\\">message.ts:10</a>)",
    }
  }
  onClick={[Function]}
/>
`;

exports[`<LogEntryMessage /> renders correctly without a source file 1`] = `
<div
  className="LogEntryMessage"
>
  <Ansi
    linkify={true}
    useClasses={true}
  >
    message
  </Ansi>
</div>
`;
